ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -889,11 +889,12 @@ func (r *RayServiceReconciler) generateConfigKeyPrefix(rayServiceInstance *rayv1
 }
 
 // Return true if healthy, otherwise false.
-func (r *RayServiceReconciler) updateAndCheckDashboardStatus(rayServiceClusterStatus *rayv1.RayServiceStatus, isHealthy bool, unhealthyThreshold *int32) bool {
+func updateAndCheckDashboardStatus(rayServiceClusterStatus *rayv1.RayServiceStatus, isHealthy bool, unhealthyThreshold *int32) bool {
 	timeNow := metav1.Now()
+	oldIsHealthy := rayServiceClusterStatus.DashboardStatus.IsHealthy
 	rayServiceClusterStatus.DashboardStatus.LastUpdateTime = &timeNow
 	rayServiceClusterStatus.DashboardStatus.IsHealthy = isHealthy
-	if rayServiceClusterStatus.DashboardStatus.HealthLastUpdateTime.IsZero() || isHealthy {
+	if rayServiceClusterStatus.DashboardStatus.HealthLastUpdateTime.IsZero() || oldIsHealthy {
```

Comment:
I don't understand the purpose of this change and how it fixes the first issue in the PR. Can you explain this more?

> KubeRay will not update the status of the RayService custom resource when the only changes are related to timestamps

It sounds like this is the real issue. Why don't we just update HealthLastUpdateTime every time? Wouldn't that fix the issue?

Reply:
> It sounds like this is the real issue. Why don't we just update HealthLastUpdateTime every time? 

This will overload the Kubernetes API Server.

Reply:
> I don't understand the purpose of this change and how it fixes the first issue in the PR. Can you explain this more?

By definition, `healthLastUpdateTime: $t` and `isHealth: true` mean that the dashboard agent is healthy from `time = $t` to `time = now`. Hence, the last second of being healthy should be the first moment we consider the dashboard agent as unhealthy.

Note that, `healthLastUpdateTime` currently does not strictly follow the definition above due to the check of head Pod status.

Reply:
Thanks, explained more offline and it makes sense. The main point I was missing is that when the health state has a transition, the status update actually goes through (there's a binary field `isHealthy` which is not a timestamp, so if it changes the status actually updates). Also to simplify the issue, the unnecessary zero-downtime upgrade bug happens any time it stays at healthy for >600s and then transitions to unhealthy; it's not specifically related to a head pod restarting.

@kevin85421 One last question.  With the new approach, if it's unhealthy for a while, then the first time it becomes healthy, we don't actually update `HealthLastUpdateTime`.  Instead, we'll update it in the next iteration.  Is this your understanding as well? I think this off-by-one issue is still okay for a hotfix.